ast/errors.go

```diff
@@ -0,0 +1,121 @@
+package ast
+
+import (
+	"strconv"
+	"strings"
+)
+
+// GetErrors will collect all of the errors from the construction of the AST.
+// If there were no errors, this will return nil. If the limit is set to
+// positive number, it will limit the number of errors to the given number. If
+// it is set to a negative number, it will return the last N errors. A value of
+// zero will return all errors.
+func GetErrors(root Node, limit int) []error {
+	v := errorVisitor{limit: limit}
+	Walk(&v, root)
+	if limit < 0 && len(v.errs) > -limit {
+		sz := -limit
+		errs := make([]error, 0, sz)
+		for i := len(v.errs) - sz; i < len(v.errs); i++ {
+			errs = append(errs, v.errs[i])
+		}
+		return errs
+	}
+	return v.errs
+}
+
+// errorDetail contains the details of the error. Err is the error itself while
+// Detail is the detailed information about the error.
+type errorDetail struct {
+	Err    string
+	Detail string
+}
+
+func (e errorDetail) Error() string {
+	var sb strings.Builder
+	sb.WriteString(e.Err)
+	if e.Detail != "" {
+		sb.WriteString(": ")
+		sb.WriteString(e.Detail)
+	}
+	return sb.String()
+}
+
+// Error is an error within the AST. It contains the information about
+// where the error happened along with the underlying cause.
+type Error struct {
+	loc   SourceLocation
+	cause errorDetail
+}
+
+func (e *Error) Error() string {
+	var sb strings.Builder
+	sb.WriteString(e.cause.Err)
+	if e.loc.Start.Line > 0 {
+		if sb.Len() > 0 {
+			sb.WriteString(":")
+		}
+		sb.WriteString(strconv.FormatInt(int64(e.loc.Start.Line), 10))
+	}
+	if e.loc.Start.Column > 0 {
+		if sb.Len() > 0 {
+			sb.WriteString(":")
+		}
+		sb.WriteString(strconv.FormatInt(int64(e.loc.Start.Column), 10))
+	}
+	if e.cause.Detail != "" {
+		if sb.Len() > 0 {
+			sb.WriteString(": ")
+		}
+		sb.WriteString(e.cause.Detail)
+	}
+	return sb.String()
+}
+
+func (e *Error) Location() SourceLocation {
+	return e.loc
+}
+
+func (e *Error) Cause() error {
+	return e.cause
+}
+
+type errorVisitor struct {
+	errs  []error
+	limit int
+}
+
+func (ev *errorVisitor) Visit(n Node) Visitor {
+	if !ev.more() {
+		return nil
+	}
+
+	// TODO(jsternberg): Fill in the details for how we retrieve errors.
+	switch n := n.(type) {
+	case *BadStatement:
```

Comment:
What is the plan for how to encode errors into the AST? This is the most important question to me.

Also remember that the AST needs to be able to serialize to JSON, which includes its errors. So whatever mechanism we choose needs to easily marshal/unmarshal to JSON.

Reply:
My initial plan would be for it to be dependent on the node itself. `BadStatement` is the easiest, but I figured that if you had an expression like `foo(a b)` it might create a property node like this:

```
{
    "type": "CallExpression",
    "callee": {
        "type": "Identifier",
        "name": "foo"
    },
    "arguments": [
        {
             "type": "ObjectExpression",
             "properties": [
                  {
                      "type": "Property",
                      "key": {
                          "type": "Identifier",
                          "name": "a"
                      },
                      "separator": {
                           "missing": true
                      },
                      "value": {
                          "type": "Identifier",
                          "name": "b"
                      }
                  }
             ]
        }
    ]
}
```

In this way, the errors are part of the contruction of the AST and don't have to be explicitly added. They are just there by fact of how it exists. It also means that the information of where the separators are is embedded as part of the AST itself for analysis programs. This is similar to how the Go AST records positions for symbols such as for [ParenExpr](https://godoc.org/go/ast#ParenExpr) where the `Lparen` and `Rparen` are part of the AST construction. In this case, we would just differentiate that it is possible for the position to be invalid and indicate that it is missing. The zero value would still be valid though and indicate that position information is missing so the errors are explicit.

Reply:
Does this mean we need to create error types for all possible syntax errors? That does not seem possible. For example how would you express this syntax error in the AST using the above plan `foo(a:)`?

Reply:
That wasn't my plan. There would be a few error nodes, such as `BadStatement` and `BadExpression` (which exist in `go/ast` too), but many of the errors would be derived from the AST that's currently produced. So the above could either be represented by having the `Value` set to `nil` (which is not allowed in a call expression context) or the value could be set to `BadExpression` with a detailed message.

This would limit the number of error AST's and I think would make the error generation more flexible. I don't think we want very many `BadXXX` types. For examples, a misformed expression such as `% a` could be represented by a `UnaryExpression` with `%` as the operator kind. Since this is not allowed, it would generate an error from the error visitor. A similar thing would be true of multiple identifiers. If we see something like `foo(a: b c)` it would be represented as the property value being a `BinaryExpression` between two expressions with no operator kind. When the error visitor sees that, it says, "well you're missing some kind of operator here".

Reply:
How are you thinking UnaryExpression will work when it has `%` as its operator? Were you thinking that this above switch statement would switch on UnaryExpressions and then check for valid operators and if it doesn't find one it would create an error? Or something else?

Reply:
Yes. I was expecting this switch statement to contain the equivalent `Validate` logic for each node, but I wanted to keep it hidden just so we had more flexibility with error handling so there wouldn't be a specific `Validate` method. I also wanted it in one location just so that we could see all error handling text in a single location rather than hunting around for the correct implementation.

Although that's an implementation detail so we could switch to using `Validate` instead pretty easily or we could switch the logic to an unexposed `validate` method. Before doing that though, I would like to experiment with this code layout.

Reply:
OK I think I understand the proposed design, but I think we are going to need a different design. Specifically we do not want the AST nodes to have behavior associated with them, rather we need them to be declarative in nature. This is because once you encode the AST to JSON and then hand that encoded JSON to an external program you can no longer leverage any behavior associated with the AST nodes, be it a GetErrors or a validate method. As such we need the errors to be explicitly encoded into the AST.

A concrete use case is Chronograf, they currently send the Flux source text to the parser via an HTTP request and get a JSON encoded AST in return. That AST needs to contain the error information. This way the Flux editor in Chronograf can communicate those errors to the user. If the error information can only be retrieved by passing the AST to a method then it is inaccessible to external programs like Chronograf. While we could extend the JSON representation of the AST to contain a list of errors how would you unmarshal that representation? Ignore the list of errors? At that point the JSON structure has duplicated information that can get out of sync. Encoding the error information directly into the AST means that the AST remains declarative (i.e. just data no behavior) and normalized (no duplicate data).

Reply:
After an offline conversation, we came to a different design. The `Errors` attribute is going to be roughly like this:

```go
package ast

// Check will inspect each node and annotate each Node with any AST errors.
func Check(n ast.Node) int

// PrintErrors will recurse through the node and print any encountered errors to the io.Writer.
// The node must have been previously compiled.
func PrintErrors(w io.Writer, n ast.Node)

type BaseNode struct {
	Loc *SourceLocation `json:"location,omitempty"`
	Errors []Error `json:"errors,omitempty"`
}

// Error is a detailed informational for the error. This exists so that the error section
// is extensible.
type Error struct {
	Message string `json:"message"`
}
```

The basic premise is that `Check` is called to determine the errors within the AST. The parser _will_ produce AST's that are invalid when given invalid input, but it will not fill out errors. This way, we can separate the logic of errors away from the parser so the parser just focuses on doing its best. It also means the logic for checking an AST can be reused for human/machine-constructed AST's and AST's constructed by the parser.

We're going to give this method an attempt to see if it works well.